jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudwatch/logs/insights/DetailedLogRecord.kt

```diff
@@ -0,0 +1,120 @@
+import software.aws.toolkits.jetbrains.utils.ApplicationThreadPoolScope
+import software.aws.toolkits.jetbrains.utils.notifyError
+import software.aws.toolkits.resources.message
+import java.lang.IllegalStateException
+import javax.swing.JButton
+import javax.swing.JPanel
+
+class DetailedLogRecord(
+    private val project: Project,
+    private val client: CloudWatchLogsClient,
+    private val logRecordPointer: String
+) : CoroutineScope by ApplicationThreadPoolScope("DetailedLogEvents"), Disposable {
+    val title = message("cloudwatch.logs.log_record", logRecordPointer)
+    lateinit var basePanel: JPanel
+    lateinit var tableView: TableView<LogRecordFieldPair>
+    private lateinit var openStream: JButton
+    private val recordLoadTask: Deferred<LogRecord>
+
+    private fun createUIComponents() {
+        val model = ListTableModel<LogRecordFieldPair>(
+            LogRecordFieldColumn(),
+            LogRecordValueColumn()
+        )
+        tableView = TableView(model).apply {
+            setPaintBusy(true)
+            emptyText.text = message("loading_resource.loading")
+        }
+        TableSpeedSearch(tableView)
+    }
+
+    init {
+        recordLoadTask = loadLogRecordAsync()
+        launch {
+            val record = recordLoadTask.await()
+            val items = record.map { it.key to it.value }
+            tableView.listTableModel.items = items
+            tableView.setPaintBusy(false)
+
+            if (items.isNotEmpty()) {
+                addOpenStreamListener(record)
+            } else {
+                tableView.emptyText.text = StatusText.DEFAULT_EMPTY_TEXT
+            }
+        }
+
+        openStream.isEnabled = false
+    }
+
+    private fun addOpenStreamListener(record: LogRecord) {
+        openStream.isEnabled = true
+        openStream.addActionListener {
+            val logGroup = record["@log"] ?: throw IllegalStateException("@log was not defined in log record")
+            val logStream = record["@logStream"] ?: throw IllegalStateException("@logStream was not defined in log record")
+
+            CloudWatchLogWindow.getInstance(project).showLogStream(
+                logGroup = getLogGroup(logGroup),
+                logStream = logStream
+            )
+        }
+    }
+
+    private fun loadLogRecordAsync(): Deferred<LogRecord> = async<LogRecord> {
+        try {
+            return@async client.getLogRecord {
+                it.logRecordPointer(logRecordPointer)
+            }.logRecord()
+        } catch (e: Exception) {
+            LOG.warn(e) { "Exception thrown while loading log record $logRecordPointer" }
+            notifyError(
+                project = project,
+                title = message("cloudwatch.logs.exception"),
+                content = ExceptionUtil.getThrowableText(e)
+            )
+        }
+
+        return@async mapOf()
+    }
+
+    override fun dispose() {
+    }
+
+    fun getComponent() = basePanel
+
+    fun isLoaded() = recordLoadTask.isCompleted
```

Comment:
This isn't strictly true (tm), obviously since this is UI code it's true enough, but why do we not set `isLoaded` in the launch? why do we need to keep track of the task as a `Deferred`?

Reply:
will add `@TestOnly` if that clears that up

Reply:
Oh then I see even less of a reason for using deferred. The comment is pretty nitpicky to begin with it's fine either way (with a `@TestOnly`.